Source/ACE/Managers/GuidManager.cs

```diff
@@ -47,29 +59,63 @@ public static class GuidManager
         // Fragmentation: None - to very light
         // We should get these on player creation.   It would be a very edge case that we threw one away.
         // Again, given probable server populations and available accounts and players - this can pose no real limitation.
-        private const uint playerMin = 0x50000001;
+        private static uint PlayerMin { get; } = 0x50000001;
 
-        private const uint playerMax = 0x5FFFFFFF;
+        private static uint PlayerMax { get; } = 0x5FFFFFFF;
+
+        private static object playerLock = new object();
 
-        // At Server startup read current max from the DB
         private static uint player = 0x50000001;
 
         // Items max ‭2,684,354,559‬
         // Many items created - relatively few persisted.   Anything that could end up in player inventory ie persisted comes from this range.
         // Fragmentation: Heavy
-        // Number Recovery Strategy - to be implemented later - but we can make is one function pull from a recycle pool that we create via 
+        // Number Recovery Strategy - to be implemented later - but we can make is one function pull from a recycle pool that we create via
         // a server operator process.   As far as the server is concerned - it just calls GetNewItemGuid.
-        private const uint itemMin = 0x6000000;
+        // TODO Fix this once we defrag.
+        // Proposed real range.
+        // private const uint itemMin = 0x6000000;
+        // private const uint itemMax = 0xFFFFFFF;
+        private const uint itemMin = 0xE0000000;
 
         private const uint itemMax = 0xFFFFFFF;
 
         // At Server startup read current max from the DB
         private static uint item = 0xDD3B018C;
 
-        public static uint NewWeenieGuid()
+        private static readonly object ShardLock = new object();
+
+        public static uint GetCurrentDbValueAsync(Action<Action<uint>> dbCall)
         {
-            weenie++;
-            return weenie;
+            uint ret = InvalidGuid;
+            dbCall.Invoke((dbVal) =>
+            {
+                lock (ShardLock)
+                {
+                    ret = dbVal;
+                    Monitor.Pulse(ShardLock);
+                }
+            });
+
+            lock (ShardLock)
+            {
+                while (ret == InvalidGuid)
+                {
+                    Monitor.Wait(ShardLock);
+                }
+            }
+
+            // NOTE: + 1 -- we don't want to reassign our current guid
+            return ret + 1;
+        }
+
+        // FIXME(ddevec): @Og -- hopefully you can use this
```

Comment:
Please change this FIXME to address that we need to do this for all initial ranges